File: exampleBots/python/bob.py
import libpyAI as ai
import math
import statistics



# main loop for this agent --- a rule-based expert system
def AI_loop():

    # release keys
    ai.thrust(0)
    ai.turnLeft(0)
    ai.turnRight(0)

    nearDanger = 170 # distance for close object (distance units)
    shotDanger = 130 # distance for close bullets (distance units)
    speedLimit = 5 
    alertDistance = 275 # backup distance threshold
    
    # handicap
    ai.setTurnSpeedDeg(20)

    # get information
    heading = int(ai.selfHeadingDeg())
    tracking = int(ai.selfTrackingDeg())

    # store in array so we can easily find the shortest
    feelers = []

    heading = int(ai.selfHeadingDeg())
    tracking = int(ai.selfTrackingDeg())
    
    frontWall = ai.wallFeeler(500, heading)
    leftWall = ai.wallFeeler(500, heading + 90)
    rightWall = ai.wallFeeler(500, heading - 90)
    trackWall = ai.wallFeeler(500, tracking)
    rearWall = ai.wallFeeler(500, heading - 180)
    backLeftWall = ai.wallFeeler(500, heading + 135)
    backRightWall = ai.wallFeeler(500, heading - 135)
    frontLeftWall = ai.wallFeeler(500, heading + 45)
    frontRightWall = ai.wallFeeler(500, heading - 45)
    
    feelers.append(frontWall)
    feelers.append(leftWall)
    feelers.append(rightWall)
    feelers.append(trackWall)
    feelers.append(rearWall)
    feelers.append(backLeftWall)
    feelers.append(backRightWall)
    feelers.append(frontLeftWall)
    feelers.append(frontRightWall)

    # collect distances
    if ai.enemyDistanceId(ai.closestShipId()) > 0:
        closestPlayerDistance = ai.enemyDistanceId(ai.closestShipId())
    else:
        closestPlayerDistance = math.inf

    if ai.shotDist(0) > 0:
        closestBulletDistance = ai.shotDist(0)
    else:
        closestBulletDistance = math.inf

    feeler = min(feelers)
    distToNearestThreat = min(closestPlayerDistance, closestBulletDistance)

    # assign priority to nearest threat
    if closestBulletDistance <= feeler and closestBulletDistance <= closestPlayerDistance: # if closest threat is a bullet
        priority = 1
    elif feeler <= closestPlayerDistance and feeler <= closestBulletDistance: # if closest threat is a wall
        priority = 2
    else: # closest threat is a player
        priority = 3

     # if the closest threat is a bullet
    if priority == 1:

        p1, p2 = (ai.selfX(), ai.selfY()), (ai.shotX(0), ai.shotY(0))
        #Compute the angle between two cartesian points, relative to horizontal
        dx = p1[0] - p2[0]
        dy = p2[1] - p1[1]
        m = -1 * (int(math.degrees(math.atan2(dy, dx))) + 180) % 360
        #Measures the difference between two headings
        m = ((m - ai.selfHeadingDeg()) + 180) % 360 - 180
        if m >= 0:
            ai.turnRight(1)
        else:
            ai.turnLeft(1)
        if ai.shotAlert(0) < shotDanger:
            ai.thrust(1)       


    # if the closest threat is a wall
    elif priority == 2:
        
        #finds difference the heading and the tracking
        head_track_diff = int(180 - abs(abs(heading - tracking) - 180))
        
        # thrust
        if ai.selfSpeed() <= speedLimit:
            ai.thrust(1)
        elif trackWall < nearDanger and head_track_diff > 90:
            ai.thrust(1)
        elif rearWall < nearDanger and head_track_diff > 90:
            ai.thrust(1)
        
        ### (production system) ###
        # turn
        if trackWall < nearDanger and leftWall < rightWall:
            ai.turnRight(1)
        elif trackWall < nearDanger and rightWall < leftWall:
            ai.turnLeft(1)
        elif backLeftWall < nearDanger and rightWall > 50:
            ai.turnRight(1)
        elif backRightWall < nearDanger and leftWall > 50:
            ai.turnLeft(1)
        elif frontRightWall < nearDanger:
            ai.turnLeft(1)
        elif frontLeftWall < nearDanger:
            ai.turnRight(1)
        elif frontWall <= alertDistance and (frontLeftWall < frontRightWall) and ai.selfSpeed() > 1: 
            ai.turnRight(1)
        elif frontWall <= alertDistance and (frontLeftWall > frontRightWall) and ai.selfSpeed() > 1:
            ai.turnLeft(1)
        elif leftWall <= alertDistance and ai.selfSpeed() > 1:
            ai.turnRight(1) 
        elif rightWall <= alertDistance and ai.selfSpeed() > 1:
            ai.turnLeft(1)
 
    # if the closest threat is a player
    elif priority == 3:
       
        p1, p2 = (ai.selfX(), ai.selfY()), (ai.shotX(0), ai.shotY(0))
        #Compute the angle between self and nearest shot, relative to horizontal
        dx = p1[0] - p2[0]
        dy = p2[1] - p1[1]
        m = -1 * (int(math.degrees(math.atan2(dy, dx))) + 180) % 360
        
        #Measures the difference between 'm' and selfs heading
        m = ((m - ai.selfHeadingDeg()) + 180) % 360 - 180
        if m <= 0:
            ai.turnRight(1)
        else:
            ai.turnLeft(1)

        p3, p4 = (ai.selfX(), ai.selfY()), (ai.screenEnemyX(0), ai.screenEnemyY(0))
        #Compute the angle between self and nearest enemy, relative to horizontal
        dx = p3[0] - p4[0]
        dy = p4[1] - p3[1]
        m2 = -1 * (int(math.degrees(math.atan2(dy, dx))) + 180) % 360
        #Measures the difference between m2 and selfs heading
        m2 = ((m2 - ai.selfHeadingDeg()) + 180) % 360 - 180
        if ai.selfHeadingDeg() <= (5 + m2) and ai.selfHeadingDeg() >= (5 - m2):
            ai.fireShot()


ai.start(AI_loop,["-name","bob"])

File: exampleBots/c/bob.c

#include <math.h>
#include <stdio.h>

AI_loop() {
    // Release keys
    thrust(0);
    turnLeft(0);
    turnRight(0);
    int speedLimit = 5;
    // Distance for close object (distance units)
    int nearDanger = 170;
    // Distance for close bullets (distance units)
    int shotDanger = 130;
    // Backup distance threshold
    int alertDistance = 275;
    // Handicap
    setTurnSpeedDeg(20);
    setPower(20);

    // Get information
    int heading = (int)selfHeadingDeg();
    int tracking = (int)selfTrackingDeg();
    // Store in array so we can easily find the shortest feeler
    int feelers[9];
    int frontWall = wallFeeler(500, heading);
    int leftWall = wallFeeler(500, heading + 90);
    int rightWall = wallFeeler(500, heading - 90);
    int trackWall = wallFeeler(500, tracking);
    int rearWall = wallFeeler(500, heading - 180);
    int backLeftWall = wallFeeler(500, heading + 135);
    int backRightWall = wallFeeler(500, heading - 135);
    int frontLeftWall = wallFeeler(500, heading + 45);
    int frontRightWall = wallFeeler(500, heading - 45);
    feelers[0] = frontWall;
    feelers[1] = leftWall;
    feelers[2] = rightWall;
    feelers[3] = trackWall;
    feelers[4] = rearWall;
    feelers[5] = backLeftWall;
    feelers[6] = backRightWall;
    feelers[7] = frontLeftWall;
    feelers[8] = frontRightWall;

    // Get distances
    int closestPlayerDistance = (int)enemyDistanceId(closestShipId());
    int closestBulletDistance = (int)shotDist(0);

    // Shortest feeler
    int feeler = feelers[0];
    for (int i = 1; i < 9; i++) {
        if (feelers[i] < feeler) {
            feeler = feelers[i];
        }
    }
    // Nearest object (player or bullet)
    int distToNearestThreat = (closestPlayerDistance < closestBulletDistance) ? closestPlayerDistance : closestBulletDistance;

    // Reset power based on how close the nearest threat is
    if (feeler <= 100 || distToNearestThreat <= 100) {
        setPower(30);
    } else if (feeler <= 50 || distToNearestThreat <= 50) {
        setPower(45);
    }

    // Assign priority to nearest threat
    int priority;
    if (closestBulletDistance <= feeler && closestBulletDistance <= closestPlayerDistance) {
        priority = 1;
    } else if (feeler <= closestPlayerDistance && feeler <= closestBulletDistance) {
        priority = 2;
    } else {
        priority = 3;
    }

    // The closest threat is a bullet
    if (priority == 1) {
        double p1[] = {selfX(), selfY()};
        double p2[] = {shotX(0), shotY(0)};
        // Get the angle between self and nearest shot, relative to horizontal
        double dx = p1[0] - p2[0];
        double dy = p2[1] - p1[1];
        int m = -1 * ((int)(atan2(dy, dx) * 180.0 / M_PI) + 180) % 360;
        // Measure the difference between 'm' and self's heading
        m = ((m - selfHeadingDeg()) + 180) % 360 - 180;

        if (m >= 0) {
            turnRight(1);
        } else {
            turnLeft(1);
        }
        if (shotAlert(0) < shotDanger) {
            thrust(1);
        }
    }
    // The closest threat is a wall
    else if (priority == 2) {
        // Find the difference between heading and tracking
        int head_track_diff = 180 - abs(abs(heading - tracking) - 180);
        // Thrusting
        if (selfSpeed() <= speedLimit) {
            thrust(1);
        } else if (trackWall < nearDanger && head_track_diff > 90) {
            thrust(1);
        } else if (rearWall < nearDanger && head_track_diff > 90) {
            thrust(1);
        }
        // Turning
        if (trackWall < nearDanger && leftWall < rightWall) {
            turnRight(1);
        } else if (trackWall < nearDanger && rightWall < leftWall) {
            turnLeft(1);
        } else if (backLeftWall < nearDanger && rightWall > 50) {
            turnRight(1);
        } else if (backRightWall < nearDanger && leftWall > 50) {
            turnLeft(1);
        } else if (frontRightWall < nearDanger) {
            turnLeft(1);
        } else if (frontLeftWall < nearDanger) {
            turnRight(1);
        } else if (frontWall <= alertDistance && (frontLeftWall < frontRightWall) && selfSpeed() > 1) {
            turnRight(1);
        } else if (frontWall <= alertDistance && (frontLeftWall > frontRightWall) && selfSpeed() > 1) {
            turnLeft(1);
        } else if (leftWall <= alertDistance && selfSpeed() > 1) {
            turnRight(1);
        } else if (rightWall <= alertDistance && selfSpeed() > 1) {
            turnLeft(1);
        }
    }
    // The closest threat is a player
    else if (priority == 3) {
        double p1[] = {selfX(), selfY()};
        double p2[] = {screenEnemyX(0), screenEnemyY(0)};
        // Get the angle between self and nearest enemy, relative to horizontal
        double dx = p1[0] - p2[0];
        double dy = p2[1] - p1[1];
        int m = -1 * ((int)(atan2(dy, dx) * 180.0 / M_PI) + 180) % 360;
        // Get the difference between m and self's heading
        m = ((m - selfHeadingDeg()) + 180) % 360 - 180;

        if (m <= 0) {
            turnRight(1);
            fireShot();
        } else {
            turnLeft(1);
            fireShot();
        }
        if (selfHeadingDeg() <= (5 + m) && selfHeadingDeg() >= (5 - m)) {
            fireShot();
        }
    }
}

int main() {
    start(AI_loop, "-name bob");
}


